src/librbd/cache/rwl/Request.h

```diff
@@ -0,0 +1,200 @@
+public:
+  using C_GuardedBlockIORequest<T>::rwl;
+
+  io::Extents image_extents;
+  bufferlist bl;
+  int fadvise_flags;
+  Context *user_req; /* User write request */
+  ExtentsSummary<io::Extents> image_extents_summary;
+  bool detained = false;                /* Detained in blockguard (overlapped with a prior IO) */
+
+  C_BlockIORequest(T &rwl, const utime_t arrived, io::Extents &&extents,
+                   bufferlist&& bl, const int fadvise_flags, Context *user_req);
+  virtual ~C_BlockIORequest();
+
+  void complete_user_request(int r);
+  void finish(int r);
+  virtual void finish_req(int r) = 0;
+
+  virtual bool alloc_resources() = 0;
+
+  void deferred();
+
+  virtual void deferred_handler() = 0;
+
+  virtual void dispatch()  = 0;
+
+  virtual const char *get_name() const override {
+    return "C_BlockIORequest";
+  }
+
+protected:
+  utime_t m_arrived_time;
+  utime_t m_allocated_time;               /* When allocation began */
+  utime_t m_dispatched_time;              /* When dispatch began */
+  utime_t m_user_req_completed_time;
+  bool m_waited_lanes = false;            /* This IO waited for free persist/replicate lanes */
+  bool m_waited_entries = false;          /* This IO waited for free log entries */
+  bool m_waited_buffers = false;          /* This IO waited for data buffers (pmemobj_reserve() failed) */
+  std::atomic<bool> m_deferred = {false}; /* Deferred because this or a prior IO had to wait for write resources */
+
+private:
+  std::atomic<bool> m_user_req_completed = {false};
+  std::atomic<bool> m_finish_called = {false};
+
+  template <typename U>
+  friend std::ostream &operator<<(std::ostream &os,
+                                  const C_BlockIORequest<U> &req);
+};
+
+struct WriteRequestResources {
+  bool allocated = false;
+  std::vector<WriteBufferAllocation> buffers;
+};
+
+/**
+ * This is the custodian of the BlockGuard cell for this write. Block
+ * guard is not released until the write persists everywhere (this is
+ * how we guarantee to each log replica that they will never see
+ * overlapping writes).
+ */
+template <typename T>
+class C_WriteRequest : public C_BlockIORequest<T> {
+public:
+  using C_BlockIORequest<T>::rwl;
+  WriteRequestResources resources;
+
+  C_WriteRequest(T &rwl, const utime_t arrived, io::Extents &&image_extents,
+                 bufferlist&& bl, const int fadvise_flags, Context *user_req);
+
+  ~C_WriteRequest();
+
+  void blockguard_acquired(GuardedRequestFunctionContext &guard_ctx);
+
+  /* Common finish to plain write and compare-and-write (if it writes) */
+  virtual void finish_req(int r);
+
+  /* Compare and write will override this */
+  virtual void update_req_stats(utime_t &now) {
+    // TODO: Add in later PRs
+  }
+  virtual bool alloc_resources() override;
+
+  /* Plain writes will allocate one buffer per request extent */
+  virtual void setup_buffer_resources(uint64_t &bytes_cached, uint64_t &bytes_dirtied);
```

Comment:
Nit: any reason why these few methods are made virtual?

Reply:
We will have C_WriteSameRequest inherited from C_WriteRequest. It has its own `setup_buffer_resources`.

Reply:
Looking at [1], it seems like it's just replacing the function? Perhaps this would be better served with an abstract base "write-op" base class?

Reply:
Like the comment about alloc_resouces  https://github.com/ceph/ceph/pull/31963#discussion_r355243192, we need to have an abstract base write class.

Reply:
Sorry, we don't need to create an abstract base write class. the behaviours of alloc_resource are same for all write requests, including normal write, writesame and compareandwrite.